doc/man7/ossl-guide-libraries-introduction.pod

```diff
@@ -0,0 +1,319 @@
+loaded with different configuration settings.
+
+If an application does not explicitly create a library context then the
+"default" library context will be used.
+
+Library contexts are represented by the B<OSSL_LIB_CTX> type. Many OpenSSL API
+functions take a library context as a parameter. Applications can always pass
+B<NULL> for this parameter to just use the default library context.
+
+The default library context is automatically created the first time it is
+needed. This will automatically load any available configuration file and will
+initialise OpenSSL for use. Unlike in earlier versions of OpenSSL (prior to
+1.1.0) no explicit initialisation steps need to be taken.
+
+Similarly when the application exits the default library context is
+automatically destroyed. No explicit de-initialisation steps need to be taken.
+
+See L<OSSL_LIB_CTX(3)> for more information about library contexts.
+See also L<ossl-guide-libcrypto-introduction(7)/ALGORITHM FETCHING>.
+
+=head1 PROPERTY QUERY STRINGS
+
+In some cases the available providers may mean that more than one implementation
+of any given algorithm might be available. For example the OpenSSL fips provider
+supplies alternative implementations of many of the same algorithms that are
+available in the OpenSSL default provider.
+
+The process of selecting an algorithm implementation is known as "fetching".
+When OpenSSL fetches an algorithm to use it is possible to specify a "property
+query string" to guide the selection process. For example a property query
+string of "provider=default" could be used to force the selection to only
+consider algorithm implementations in the default provider.
+
+Property query strings can be specified explicitly as an argument to a function.
+It is also possible to specify a default property query string for the whole
+library context using the L<EVP_set_default_properties(3)> or
+L<EVP_default_properties_enable_fips(3)> functions. Where both
+default properties and function specific properties are specified then they are
+combined. Function specific properties will override default properties where
+there is a conflict.
+
+See L<ossl-guide-libcrypto-introduction(7)/ALGORITHM FETCHING> for more
+information about fetching. See L<property(7)> for more information about
+properties.
+
+=head1 MULTI-THREADED APPLICATIONS
+
+As long as OpenSSL has been built with support for threads (the default case
+on most platforms) then most OpenSSL I<functions> are thread-safe in the sense
+that it is safe to call the same function from multiple threads at the same
+time. However most OpenSSL I<data structures> are not thread-safe. For example
+the L<BIO_write(3)> and L<BIO_read(3)> functions are thread safe. However it
+would not be thread safe to call BIO_write() from one thread while calling
+BIO_read() in another where both functions are passed the same B<BIO> object
+since both of them may attempt to make changes to the same B<BIO> object.
```

Comment:
BIO_read() can do changes to passed object? I mean this may be true, it's just me, who doesn't know BIO_read() details. It is strange for me to read "both of them may attempt to make changes"... I wouldn't expect changes from *_read().

Reply:
Many of our BIOs will make changes on a read. For example to buffer data.